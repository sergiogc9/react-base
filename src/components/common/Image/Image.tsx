import React from 'react';
import { useUpdateEffect } from '@src/lib/hooks';

type ComponentProps = {
	url: string
	fallbackUrl?: string
	fallbackComponent?: JSX.Element
};

const Image: React.FunctionComponent<ComponentProps> = React.memo(props => {
	const { url, fallbackUrl, fallbackComponent, children } = props;
	const [loadFailed, setLoadFailed] = React.useState(false);
	useUpdateEffect(() => {
		if (url) return setLoadFailed(false);
	}, [url]);

	const showFallback = loadFailed || url === undefined;
	const isFallbackComponent = showFallback && fallbackComponent;
	const isChildren = showFallback && children;
	if (showFallback && fallbackUrl) return <img src={fallbackUrl} alt="" />;
	if (isFallbackComponent || isChildren) return fallbackComponent! || children;

	return <img src={url} onError={e => setLoadFailed(true)} alt="" />;
});

export default Image;
