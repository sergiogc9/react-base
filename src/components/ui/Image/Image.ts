import { Flex, FlexProps } from '@sergiogc9/react-ui';

const Image: React.FC<FlexProps<React.ImgHTMLAttributes<HTMLImageElement>>> = Flex;

export default Image;
