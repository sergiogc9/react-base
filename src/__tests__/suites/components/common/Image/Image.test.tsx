import React from "react";
import { mount } from "enzyme";
import { act } from 'react-dom/test-utils';

import Image from '@src/components/common/Image/Image';

let setLoadFailed = jest.fn();

const mockSetLoad = (loaded: boolean) => {
	jest
		.spyOn(React, 'useState')
		.mockImplementation(() => [loaded, setLoadFailed]);
};

describe("Image common component", () => {
	let wrapper: any;
	const getComponent = () => (
		<Image
			url="http://fakeurl.com"
			fallbackUrl="http://fallback.fake.url"
		/>
	);

	const shallowWrapper = (component?: JSX.Element) => {
		act(() => { wrapper = mount(component || getComponent()); });
	};

	beforeEach(() => {
		setLoadFailed = jest.fn();
		mockSetLoad(false);
		shallowWrapper();
	});

	it("Component default snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with fallbackUrl", () => {
		mockSetLoad(true);
		shallowWrapper();
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with component fallback", () => {
		mockSetLoad(true);
		wrapper.setProps({ fallbackUrl: undefined, fallbackComponent: <div>Fallback component</div> });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with component fallback in children", () => {
		mockSetLoad(true);
		shallowWrapper(<Image url="http://fakeurl.com"><div>Fallback component in children</div></Image>);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Image url handler modifies state", () => {
		wrapper.find('img').props().onError({ target: {} });
		expect(setLoadFailed).toHaveBeenCalledWith(true);
	});

	it("loadFailed set as false when modifying url prop", () => {
		mockSetLoad(true);
		wrapper.setProps({ url: "http://another-fake.url" });
		expect(setLoadFailed).toHaveBeenCalledWith(false);
		wrapper.unmount();
	});
});
