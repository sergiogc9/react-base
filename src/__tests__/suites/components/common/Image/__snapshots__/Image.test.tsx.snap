// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image common component Component default snapshot 1`] = `"<img src=\\"http://fakeurl.com\\" alt=\\"\\">"`;

exports[`Image common component Component snapshot with component fallback 1`] = `"<div>Fallback component</div>"`;

exports[`Image common component Component snapshot with component fallback in children 1`] = `"<div>Fallback component in children</div>"`;

exports[`Image common component Component snapshot with fallbackUrl 1`] = `"<img src=\\"http://fallback.fake.url\\" alt=\\"\\">"`;
